clusterloader2/pkg/measurement/common/probes/probes.go

```diff
@@ -0,0 +1,338 @@
+
+	checkProbesReadyInterval = 30 * time.Second
+	checkProbesReadyTimeout  = 5 * time.Minute
+
+	currentProbesMetricsVersion = "v1"
+)
+
+var probesEnabled bool
+
+func init() {
+	flags.BoolEnvVar(&probesEnabled, "enable-probes", "ENABLE_PROBES", false, "Whether to run the probes in the clusterloader2 tests.")
+	measurement.Register(name, createProbesMeasurement)
+}
+
+func createProbesMeasurement() measurement.Measurement {
+	return &probesMeasurement{
+		probeNameToPrometheusQueryTmpl: map[string]string{
+			"in_cluster_network_latency": "quantile_over_time(0.99, probes:in_cluster_network_latency:histogram_quantile[%v])",
+		},
+	}
+}
+
+type probesMeasurement struct {
+	// probeNameToPrometheusQueryTmpl defines a config of this measurement. Updating the config in the
+	// createProbesMeasurement method is the only place in go code that needs to be changed while
+	// adding a new probe.
+	// Each query template should accept a single %v placeholder corresponding to the query window
+	// length. See the 'in_cluster_network_latency' as an example.
+	probeNameToPrometheusQueryTmpl map[string]string
+
+	framework        *framework.Framework
+	replicasPerProbe int
+	templateMapping  map[string]interface{}
+	nServiceMonitors int
+	startTime        time.Time
+	mutex            sync.Mutex
+}
+
+// Execute supports two actions:
+// - start - starts probes and sets up monitoring
+// - gather - Gathers and prints metrics.
+func (p *probesMeasurement) Execute(config *measurement.MeasurementConfig) ([]measurement.Summary, error) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	if !probesEnabled {
+		klog.Info("Probes are disabled, skipping the measurement execution")
+		return nil, nil
+	}
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+	switch action {
+	case "start":
+		klog.Info("Starting probes...")
+		err := p.start(config)
+		return nil, err
+	case "gather":
+		klog.Info("Gathering metrics from probes...")
+		summary, err := p.gather(config.Params)
+		if err != nil && !errors.IsMetricViolationError(err) {
+			return nil, err
+		}
+		return []measurement.Summary{summary}, nil
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+// Dispose cleans up after the measurement.
+func (p *probesMeasurement) Dispose() {
+	if !probesEnabled {
+		klog.Info("Probes are disabled, skipping the Dispose() step")
+		return
+	}
+	klog.Info("Stopping probes...")
+	k8sClient := p.framework.GetClientSets().GetClient()
+	if err := client.DeleteNamespace(k8sClient, probesNamespace); err != nil {
+		klog.Errorf("error while deleting %s namespace: %v", probesNamespace, err)
+	}
+	if err := client.WaitForDeleteNamespace(k8sClient, probesNamespace); err != nil {
+		klog.Errorf("error while waiting for %s namespace to be deleted: %v", probesNamespace, err)
+	}
+	*p = *createProbesMeasurement().(*probesMeasurement) // Reset measurement state.
```

Comment:
I don't think that's needed - we don't need to reinitialize it becasue the same measurement instantiation won't be reused.